### PR TITLE
Database types boolean optional

### PR DESCRIPTION
**Pull request by jodosha**
Let database to raise an error when 'NOT NULL' constraint is violated for column type boolean and `nil` value.

This is part of the integration with `hanami-model`.

/cc @solnic

### CODE REVIEW

**Comment by flash-gordon** [comment]
hey @jodosha 
I think we need a slightly different type for this
I tested locally this one:
```ruby
      BoolType = ::Class.new do
        include Dry::Types::Builder
        include Dry::Types::Options
        include Dry::Equalizer(:options)

        def valid?(value)
          value == true || value == false
        end

        def optional
          Types::Optional::Bool
        end

        def default?
          false
        end

        def call(value)
          value
        end
      end

      Bool = BoolType.new
```
hanami-model works fine with that^
@solnic wdyt?
I guess we need support for such dummy types in dry-t

**Comment by jodosha** [comment]
@flash-gordon Thanks for the quick reply. I'm fine with it as long both `rom-sql` and `hanami-model` work fine. But what's the difference with `Bool   = ROM::Types::Bool.optional`?

**Comment by flash-gordon** [comment]
`ROM::Types::Bool.optional` allows only `nil`, `true`, and `false`, but we decided not to check types by default and this means `optional` is not consistent with other types

**Comment by jodosha** [comment]
@flash-gordon I see. What's a good way to test the extra behavior that the implementation using `BoolType` would introduce?

**Comment by flash-gordon** [comment]
well, I think we can pass something meaningless, e.g. `1`, and test if error occurs in DB-adapter

**Comment by jodosha** [comment]
@flash-gordon With both the implementations it fails with:

```shell
1) Commands / Create with postgres re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 1) }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: PG::DatatypeMismatch: ERROR:  column "cute" is of type boolean but express...                                         ^
       HINT:  You will need to rewrite or cast the expression.
       > with backtrace:
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `async_exec'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `block in execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:149:in `check_disconnect_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:526:in `_execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block (2 levels) in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:547:in `check_database_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `block in synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:970:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:676:in `fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:1078:in `returning_fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:336:in `insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/shared/postgres.rb:1417:in `insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:14:in `block in insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `map'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `insert'
         # ./lib/rom/sql/commands/create.rb:32:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command.rb:110:in `call'
         # ./lib/rom/sql/commands/error_wrapper.rb:12:in `call'
         # ./spec/integration/commands/create_spec.rb:202:in `block (5 levels) in <top (required)>'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command_registry.rb:57:in `try'
         # ./spec/integration/commands/create_spec.rb:202:in `block (4 levels) in <top (required)>'
         # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  2) Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 1) }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError but nothing was raised
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  3) Commands / Create with mysql re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 1) }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError but nothing was raised
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

Finished in 8.95 seconds (files took 0.56113 seconds to load)
63 examples, 3 failures

Failed examples:

rspec './spec/integration/commands/create_spec.rb[1:2:7]' # Commands / Create with postgres re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:1:7]' # Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:3:7]' # Commands / Create with mysql re-raises not-null constraint violation error with wrong type
```

**Comment by flash-gordon** [comment]
wow :) what about something _even more_ meaningless? 😆  Don't know, `Date`? We definitely should get an error on this :)

**Comment by jodosha** [comment]
@flash-gordon Nope, I tried with `Date` and a gibberish `String`. With this attempt PG and MySQL are complaining with an error about wrong database type.

**Comment by flash-gordon** [comment]
@jodosha so, invalid error type is expected, we should test it thrown. `BoolType` is anaemic, it does not touch value, hence it won't raise an error

**Comment by jodosha** [comment]
@flash-gordon you're right, but this isn't probably the right test to write as I get the same result with both the implementations.

**Comment by flash-gordon** [comment]
@jodosha how so? Using `ROM::Types::Bool.optional` should end up with `Dry::Types::SchemaError` and using `BoolType` should raise kinda TypeMismatch error

**Comment by jodosha** [comment]
@flash-gordon `¯\_(ツ)_/¯`

With the current implementation:

```shell
➜ be rspec spec/integration/commands/create_spec.rb

Randomized with seed 55431
F........................F...................F.................

Failures:

  1) Commands / Create with postgres re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: PG::InvalidTextRepresentation: ERROR:  invalid input syntax for type boole...") VALUES ('Charlie', 'you can b...
                                                                    ^
       > with backtrace:
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `async_exec'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `block in execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:149:in `check_disconnect_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:526:in `_execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block (2 levels) in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:547:in `check_database_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `block in synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:970:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:676:in `fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:1078:in `returning_fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:336:in `insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/shared/postgres.rb:1417:in `insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:14:in `block in insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `map'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `insert'
         # ./lib/rom/sql/commands/create.rb:32:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command.rb:110:in `call'
         # ./lib/rom/sql/commands/error_wrapper.rb:12:in `call'
         # ./spec/integration/commands/create_spec.rb:202:in `block (5 levels) in <top (required)>'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command_registry.rb:57:in `try'
         # ./spec/integration/commands/create_spec.rb:202:in `block (4 levels) in <top (required)>'
         # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  2) Commands / Create with mysql re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: Mysql2::Error: Incorrect integer value: 'you can bet' for column 'cute' at row 1> with backtrace:
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:120:in `_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:120:in `block in query'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:119:in `handle_interrupt'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:119:in `query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:137:in `block in _execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:132:in `_execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/utils/mysql_mysql2.rb:36:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `block in synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/utils/mysql_mysql2.rb:36:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:76:in `execute_insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:986:in `execute_insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:338:in `insert'
         # ./lib/rom/sql/relation/writing.rb:37:in `insert'
         # ./lib/rom/sql/commands/create.rb:42:in `block in insert'
         # ./lib/rom/sql/commands/create.rb:42:in `map'
         # ./lib/rom/sql/commands/create.rb:42:in `insert'
         # ./lib/rom/sql/commands/create.rb:32:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command.rb:110:in `call'
         # ./lib/rom/sql/commands/error_wrapper.rb:12:in `call'
         # ./spec/integration/commands/create_spec.rb:202:in `block (5 levels) in <top (required)>'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command_registry.rb:57:in `try'
         # ./spec/integration/commands/create_spec.rb:202:in `block (4 levels) in <top (required)>'
         # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  3) Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError but nothing was raised
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

Finished in 9.17 seconds (files took 0.43393 seconds to load)
63 examples, 3 failures

Failed examples:

rspec './spec/integration/commands/create_spec.rb[1:2:7]' # Commands / Create with postgres re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:3:7]' # Commands / Create with mysql re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:1:7]' # Commands / Create with sqlite re-raises not-null constraint violation error with wrong type

Randomized with seed 55431
```

---

With `BoolType`

```shell
➜ be rspec spec/integration/commands/create_spec.rb

Randomized with seed 13453
......F...................F....................F...............

Failures:

  1) Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError but nothing was raised
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  2) Commands / Create with postgres re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: PG::InvalidTextRepresentation: ERROR:  invalid input syntax for type boole...") VALUES ('Charlie', 'you can b...
                                                                    ^
       > with backtrace:
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `async_exec'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `block in execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:186:in `execute_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:149:in `check_disconnect_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:173:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:526:in `_execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block (2 levels) in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:547:in `check_database_errors'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `block in synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:342:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:970:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/postgres.rb:676:in `fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:1078:in `returning_fetch_rows'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:336:in `insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/shared/postgres.rb:1417:in `insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:14:in `block in insert'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `map'
         # ./lib/rom/sql/extensions/postgres/commands.rb:13:in `insert'
         # ./lib/rom/sql/commands/create.rb:32:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command.rb:110:in `call'
         # ./lib/rom/sql/commands/error_wrapper.rb:12:in `call'
         # ./spec/integration/commands/create_spec.rb:202:in `block (5 levels) in <top (required)>'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command_registry.rb:57:in `try'
         # ./spec/integration/commands/create_spec.rb:202:in `block (4 levels) in <top (required)>'
         # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

  3) Commands / Create with mysql re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: Mysql2::Error: Incorrect integer value: 'you can bet' for column 'cute' at row 1> with backtrace:
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:120:in `_query'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:120:in `block in query'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:119:in `handle_interrupt'
         # /Users/luca/.gem/ruby/2.3.1/gems/mysql2-0.4.5/lib/mysql2/client.rb:119:in `query'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:137:in `block in _execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:132:in `_execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/utils/mysql_mysql2.rb:36:in `block in execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `block in synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/database/connecting.rb:280:in `synchronize'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/utils/mysql_mysql2.rb:36:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/adapters/mysql2.rb:76:in `execute_insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:986:in `execute_insert'
         # /Users/luca/.gem/ruby/2.3.1/gems/sequel-4.40.0/lib/sequel/dataset/actions.rb:338:in `insert'
         # ./lib/rom/sql/relation/writing.rb:37:in `insert'
         # ./lib/rom/sql/commands/create.rb:42:in `block in insert'
         # ./lib/rom/sql/commands/create.rb:42:in `map'
         # ./lib/rom/sql/commands/create.rb:42:in `insert'
         # ./lib/rom/sql/commands/create.rb:32:in `execute'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command.rb:110:in `call'
         # ./lib/rom/sql/commands/error_wrapper.rb:12:in `call'
         # ./spec/integration/commands/create_spec.rb:202:in `block (5 levels) in <top (required)>'
         # /Users/luca/.gem/ruby/2.3.1/bundler/gems/rom-179772b11bd3/lib/rom/command_registry.rb:57:in `try'
         # ./spec/integration/commands/create_spec.rb:202:in `block (4 levels) in <top (required)>'
         # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'
     # ./spec/integration/commands/create_spec.rb:201:in `block (3 levels) in <top (required)>'

Finished in 8.51 seconds (files took 0.45353 seconds to load)
63 examples, 3 failures

Failed examples:

rspec './spec/integration/commands/create_spec.rb[1:1:7]' # Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:2:7]' # Commands / Create with postgres re-raises not-null constraint violation error with wrong type
rspec './spec/integration/commands/create_spec.rb[1:3:7]' # Commands / Create with mysql re-raises not-null constraint violation error with wrong type

Randomized with seed 13453
```

**Comment by flash-gordon** [comment]
@jodosha quite surprising :) I'll have a thorough look in a few hours

**Comment by flash-gordon** [comment]
@jodosha aha, that's because you didn't inferred schema for your relation! :)
see specs I wrote for inserts: https://github.com/rom-rb/rom-sql/blob/master/spec/integration/schema_inference_spec.rb#L97

**Comment by jodosha** [comment]
@flash-gordon Thanks for spotting the problem. Please have a look at 07baf6ca610d04b52e74fd18d3cb83def999bc38 I added it in the before block.

---

Now if we add a new test:

```ruby
    it 're-raises not-null constraint violation error with wrong type' do
      puppies = commands[:puppies]

      expect {
        puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
      }.to raise_error(ROM::SQL::NotNullConstraintError)
    end
```

---

With the current implementation we get the same error for **all the three** databases:

```shell
1) Commands / Create with mysql re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<Dry::Types::SchemaError: "you can bet" (String) has invalid type for :cute> with backtrace:
# ...
```

---

With `BoolType` we get the same errors reported in my last comment:

```shell
1) Commands / Create with postgres re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: PG::InvalidTextRepresentation: ERROR:  invalid input syntax for type boole...") VALUES ('Charlie', 'you can b...
                                                                    ^
       > with backtrace:

# ...

 2) Commands / Create with mysql re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError, got #<ROM::SQL::DatabaseError: Mysql2::Error: Incorrect integer value: 'you can bet' for column 'cute' at row 1> with backtrace:

# ...

 3) Commands / Create with sqlite re-raises not-null constraint violation error with wrong type
     Failure/Error:
       expect {
         puppies.try { puppies.create.call(name: 'Charlie', cute: 'you can bet') }
       }.to raise_error(ROM::SQL::NotNullConstraintError)

       expected ROM::SQL::NotNullConstraintError but nothing was raised
     # ./spec/integration/commands/create_spec.rb:205:in `block (3 levels) in <top (required)>'

# ...
```

---

WDYT? Which behavior do we want to expose?

**Comment by flash-gordon** [comment]
The last one, `DatabaseError` is what we're expecting when we use anaemic types. If you do want to check types or coerce values manually, you need to use use manually described schema. However, I think adding `strict` option to the inferrer (with false default value) is a good idea fwiw

**Comment by jodosha** [comment]
@flash-gordon Thank you very much. I think that probably we shouldn't be "_smart_" and take a decision on missing/nil values. It could lead to subtle bugs.

But because this is ROM and not Hanami the final decision it's up to you folks. @solnic WDYT?

File: spec/unit/types_spec.rb [hunk 1]
```diff
@@ -11,6 +11,29 @@
...
+    end
+
+    it 'returns false when false is given' do
+      input  = false
+      output = described_class[input]
+
+      expect(output).to be(false)
+    end
+  end
+
```
Comment (solnic) [inline spec/unit/types_spec.rb]: And this can be removed too

**Comment by solnic** [comment]
Please add `dry-types` from master in this branch, and your specs should all pass

**Comment by jodosha** [comment]
@solnic @flash-gordon Thank you folks. I think this should be ready to be merged. 😄

**Comment by solnic** [comment]
I cleaned up commits so only relevant ones were merged into master. Closing this one. Pheew :)

**Pull request closed without merging**